gunpowder/nodes/buffered_predict.py

```diff
@@ -0,0 +1,356 @@
+import ctypes
+import logging
+import multiprocessing as mp
+import numpy as np
+
+from functools import reduce
+from gunpowder.array import ArrayKey, Array
+from gunpowder.nodes.generic_predict import GenericPredict
+from operator import mul
+
+logger = logging.getLogger(__name__)
+
+
+class BufferedPredict(GenericPredict):
+    '''Buffered implementation of :class:`gunpowder.nodes.Predict`.
+
+    The class runs a background predict process with shared memory buffer
```

Comment:
Not quite sure if I get how the BufferedPredict differs from the `spawn_subprocess` flag in GenericPredict.
I may be confused on how the existing predict node works, but I thought the point with that flag was to provide a gpu worker in a seperate process so that we could have multiple workers generating batches and feed them to the predict process. Is the difference just the use of a shared memory buffer instead of the multiprocessing.queue to pass the inputs to the gpu process?

Reply:
Yea there are a few problems with the current implementation of GenericPredict. First is that it does not allow for multiple workers ([number of workers is always 1](https://github.com/funkey/gunpowder/blob/v1.3-dev/gunpowder/nodes/generic_predict.py#L100)). Increasing the number of workers will lead to some complicated queue reordering problem without locking (worker A/B could get the results B/A from the background predict process). This leads to the current implementation, when using with [daisy predict node](https://github.com/funkey/gunpowder/blob/v1.3-dev/gunpowder/nodes/daisy_request_blocks.py), that spawns a copy of the network model, each with its own weight/memory for each worker which can quickly lead to out of memory problem with multiple workers especially for big models. Also, `multiprocessing.queue` uses `pickle` to serialize data first before queuing which is quite a bit slower than using shared memory.

In short BufferedPredict solves 2 main problems with GenericPredict.
1. Low performance of `multiprocessing.queue`
2. Replicated memory usage of multiple workers

Reply:
I'm surprised that the daisy node + Predict creates separate copies of the model per cpu worker. I don't have an example on me right now but I'm pretty sure I've used these nodes with many workers in the past and it didn't run out of memory. If multiple copies of the model were initialized on the gpu then I wouldn't expect this combo to ever work with more than just a few workers. Also I think setup() should also only be called once since the workers in daisy request blocks reuse the same [upstream provider](https://github.com/funkey/gunpowder/blob/4c115bbd944fdc1929427692a9742709c86c12de/gunpowder/nodes/daisy_request_blocks.py#L133), although here I'm getting a bit further into the details of multiprocessing than I am familiar with and I'm not entirely sure what is referenced and what is copied when creating workers in separate threads/processes.

Improving the performance of the multiprocessing.queue sounds good to me. Do we need a separate node for this? If it reduces memory overhead and speeds up data transfer, it seems like a win win and should just replace the old `GenericPredict` behavior.

Reply:
Ah, I think I must have used the Torch predict node without `spawn_subprocess` which may have spawned a copy for each worker. I'll try to rebase the jax predict node to `GenericPredict`. If it works it might be a cleaner add for now and we can work on the shared memory queue later.

Regarding shared memory queue, I'm a bit hesitant replacing multiprocessing.queue completely because it is not as flexible. Currently it only supports transferring shared np.ndarray and pretty much nothing else, so it's probably sufficient for production deployment but not for research/more experimental uses.

Reply:
I rebased jax predict to GenericPredict and added some test cases (based on the PyTorch ones). For some reason I couldn't get the multiprocessing tests to not hang with the `Scan` and `PreCache` nodes... maybe there's incompatibilities with jax and multiprocessing and pytest (`PreCache` and `DaisyRequestBlocks` worked fine outside of tests).